### PR TITLE
add new learing rate strategy to reduce lr when loss reach on plateau

### PR DESCRIPTION
**Pull request by zhwesky2010**
**添加loss自适应的学习率衰减策略。**

--- 

image
image
image

### CODE REVIEW

File: python/paddle/fluid/dygraph/learning_rate_scheduler.py [hunk 1]
```diff
@@ -619,7 +622,7 @@ class LinearLrWarmup(LearningRateDecay):
     
         learning_rate = 0.1 
         warmup_steps = 50
-        start_lr = 1. / 3.
+        start_lr = 0
```
Comment (phlrain) [inline python/paddle/fluid/dygraph/learning_rate_scheduler.py:625]: 为什么要改这个 LinearWarmup，是之前有bug么

File: python/paddle/fluid/layers/learning_rate_scheduler.py [hunk 2]
```diff
@@ -555,3 +566,194 @@ def linear_lr_warmup(learning_rate, warmup_steps, start_lr, end_lr):
                             shape=[1], dtype=dtype, value=float(learning_rate))
                     tensor.assign(learning_rate, lr)
             return lr
+
+
+def reduce_lr_on_plateau(indicator,
```
Comment (phlrain) [inline python/paddle/fluid/layers/learning_rate_scheduler.py]: 这个地方为什么不直接叫loss

File: python/paddle/fluid/dygraph/learning_rate_scheduler.py (hunk 1 above)
Comment (zhwesky2010) [reply python/paddle/fluid/dygraph/learning_rate_scheduler.py:625]: > 为什么要改这个 LinearWarmup，是之前有bug么

这个示例代码之前有bug，sample code跑的失败了，start_lr必须 < end_lr

File: python/paddle/fluid/layers/learning_rate_scheduler.py (hunk 2 above)
Comment (zhwesky2010) [reply python/paddle/fluid/layers/learning_rate_scheduler.py]: > 这个地方为什么不直接叫loss

也可以，不过tensorflow是叫monitor，pytorch叫metric

File: python/paddle/fluid/dygraph/learning_rate_scheduler.py [hunk 3]
```diff
@@ -660,3 +663,153 @@ def step(self):
...
+    def step(self, loss):
+        # loss must be 1-D Tensor with shape [1]
+        check_type(loss, 'loss', Variable, 'ReduceLROnPlateau.step')
+        assert len(loss.shape) == 1 and loss.shape[0] == 1, "the loss.shape " \
+            "should be (1L,), but the current loss.shape is {}. Maybe that "  \
+            "you should call fluid.layers.mean to process it first.".format(loss.shape)
+
+        if self.cooldown_counter > 0:
+            self.cooldown_counter -= 1
+            return self.learning_rate
```
Comment (phlrain) [inline python/paddle/fluid/dygraph/learning_rate_scheduler.py]: 这个step 需要返回值么？  还是只要把 self.learning_rate的值设置好了就行了

Comment (zhwesky2010) [reply python/paddle/fluid/dygraph/learning_rate_scheduler.py]: step不用返回值

**Review by phlrain — DISMISSED** [review]

File: python/paddle/fluid/dygraph/learning_rate_scheduler.py [hunk 4]
```diff
@@ -660,3 +663,191 @@ def step(self):
...
+                 patience=10,
+                 verbose=False,
+                 threshold=1e-4,
+                 threshold_mode='rel',
+                 cooldown=0,
+                 min_lr=0,
+                 eps=1e-8,
+                 dtype='float32'):
+        super(ReduceLROnPlateau, self).__init__(dtype=dtype)
+        if mode not in ['min', 'max']:
```
Comment (lanxianghit) [inline python/paddle/fluid/dygraph/learning_rate_scheduler.py:759]: 对str类型的参数应该能兼容各种大小写形式，比如min，MIN，Min等，建议先做lowercase处理。

Comment (lanxianghit) [reply python/paddle/fluid/dygraph/learning_rate_scheduler.py:759]: 由于发版时间限制，可以先merge，后续再修改

**Review by lanxianghit — DISMISSED** [review]

File: python/paddle/fluid/dygraph/learning_rate_scheduler.py (hunk 4 above)
Comment (zhwesky2010) [reply python/paddle/fluid/dygraph/learning_rate_scheduler.py:759]: OK

**Review by lanxianghit — APPROVED** [review]

**Review by saxon-zh — APPROVED** [review]
LGTM

File: python/paddle/fluid/dygraph/learning_rate_scheduler.py (hunk 4 above)
Comment (zhwesky2010) [reply python/paddle/fluid/dygraph/learning_rate_scheduler.py:759]: 已修改
